File: main/scala/com/awesomesauce/minecraft/forge/openautomation/api/lasers/LaserAPI.scala
package com.awesomesauce.minecraft.forge.openautomation.api.lasers

import net.minecraft.tileentity.TileEntity
import net.minecraft.world.World
import net.minecraftforge.common.util.ForgeDirection

import scala.collection.mutable.ArrayBuffer

object NullCallback extends LaserCallback {
  def getName = "Nothing"

  def getDescription = "Do Nothing."

  def isUseableOn(destWorld: World, destX: Int, destY: Int, destZ: Int, destTo: ForgeDirection): Boolean = true

  def isExecutable(callbackEmitter: TileEntity, destWorld: World, destX: Int, destY: Int, destZ: Int, destTo: ForgeDirection): Boolean = true

  def executeCallback(callbackEmitter: TileEntity, destWorld: World, destX: Int, destY: Int, destZ: Int, destTo: ForgeDirection) = {}
}

object LaserAPI {
  val callbacks: ArrayBuffer[LaserCallback] = ArrayBuffer[LaserCallback](NullCallback)

  def registerCallback(callback: LaserCallback) = {
    callbacks.append(callback)
  }
}

File: main/java/com/awesomesauce/minecraft/forge/openautomation/api/lasers/LaserAPI.java
package com.awesomesauce.minecraft.forge.openautomation.api.lasers;

import net.minecraft.item.ItemStack;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class LaserAPI {
    public static List<LaserCallback> callbacks = new ArrayList<LaserCallback>();

    public static void registerCallback(LaserCallback callback) {
        callbacks.add(callback);
    }

    public static Set<ILaserModule> modules = new HashSet<ILaserModule>();

    public static void registerLaserModule(ILaserModule module) {
        modules.add(module);
    }

    public static boolean isModule(ItemStack stack) {
        for (ILaserModule module : modules) {
            if (module.stackIsModule(stack))
                return true;
        }
        return false;
    }

    public static ILaserModule getModule(ItemStack stack) {
        for (ILaserModule module : modules) {
            if (module.stackIsModule(stack))
                return module;
        }
        return null;
    }
}
